kubeai: document usage of NRI plugins for performance optimization (#1113)
opea-project/GenAIInfra: kubeai/README.md

Enjoy the answer!

# CPU Performance Optimization with NRI

## Overview

[NRI plugins][nri-plugins] provide a way to
optimize the resource placement of applications in a Kubernetes cluster. They
connect to the container runtime and are able, for example, to adjust the CPU
and memory pinning of containers.

This section provides a guide on how to use the
[Balloons Policy][balloons-policy] plugin from the [NRI Plugins][nri-plugins]
project to optimize the performance of CPU-backed KubeAI profiles.

## Installation of Balloons Policy Plugin

> > **NOTE:** In order to prevent disturbing already running workloads it is
> > recommended to install the NRI plugin to an empty node (right after node
> > bootstrap or drain the node before installation).

Reviewer: ```suggestion
> **NOTE:** To avoid disturbing already running workloads it is
> recommended to install the NRI plugin to an empty node (do it right after node
> bootstrap, or to drain the node before installation).
```

Author: Thx, fixed

Install the balloons policy plugin with Helm:

```bash
helm repo add nri-plugins https://containers.github.io/nri-plugins
helm repo update nri-plugins
helm install -n kube-system balloons nri-plugins/nri-resource-policy-balloons
```

> > **NOTE**: If using containerd version earlier than v2.0 you need to enable
> > the NRI support in the containerd configuration file. Instead of manual
> > configuration you can provide `--set nri.runtime.patchConfig=true` to the Helm
> > command above, which will automatically patch the containerd configuration
> > file on each node.

Reviewer: ```suggestion
> **NOTE**: With `containerd` version earlier than v2.0 you need to enable
> the NRI support in the `containerd` configuration file. Instead of manual
> configuration you can provide `--set nri.runtime.patchConfig=true` to the Helm
> command above, which will automatically patch the `containerd` configuration
> file on each node.
```

Author: Thx. Did s/If using/With/.

However I didn't put containerd in backticks as referring to the containerd project/package rather than a shell command directly. Dunno if there is some OPEA style guide, probably not :smile:  Thoughts?

Reviewer 2: this is fine.

Reviewer: Ah, it's officially written with lower-case `c`.  One option would be to use italics for names, but this is fine as-is.

Verify that the balloons policy plugin is running on every node:

```bash
$ kubectl -n kube-system get ds nri-resource-policy-balloons
NAME                           DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
nri-resource-policy-balloons   2         2         2       2            2           kubernetes.io/os=linux   77s
```

## Configuration of Balloons Policy Plugin

The aim of the balloons policy configuration is to isolate the vllm containers

> to minimize their interference with other containers.

Author: Thx, changed. Except for vllm as now I'm referring to inference in more generic terms (added support for ollama, too)

An example configuration for the current CPU-backed model profiles:

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: config.nri/v1alpha1
kind: BalloonsPolicy
metadata:
  name: default
  namespace: kube-system
spec:
  allocatorTopologyBalancing: true
  balloonTypes:
  - name: kubeai-vllm
    allocatorPriority: high
    minBalloons: 1
    minCPUs: 12
    preferNewBalloons: true
    hideHyperthreads: true
    matchExpressions:
    - key: labels/app.kubernetes.io/name
      operator: In
      values: ["vllm"]
    - key: name
      operator: In
      values: ["server"]
  - name: default
    namespaces:
    - "*"
    shareIdleCPUsInSame: numa
  pinCPU: true
  pinMemory: false
  reservedPoolNamespaces:
  - kube-system
  reservedResources:
    cpu: "2"
  log:
    debug: ["policy"]
  instrumentation:
    httpEndpoint: :8891
    prometheusExport: true
EOF
```

The configuration above allocates at minimum 12 logical CPUs (6 full CPU cores

> assuming that hyperthreading is enabled) for the inference serving pods. With

Reviewer: Pods, not containers?
```suggestion
assuming that hyperthreading is enabled) for the vLLM pods. With
```

Author: I heavily rephrased this paragraph. No reference to vLLM anymore (just "inference")

the settings of the `kubeai-vllm` balloon it creates a new dedicated balloon

> for each inference engine container as long as there are enough free cpus on

Reviewer: ```suggestion
for each inference engine container as long as there are enough free CPUs on
```

Author: Fixed

> install the NRI plugin to an empty node (do it right after node bootstrap, or
> drain the node before installation).

Install the balloons policy plugin with Helm:

```bash
helm repo add nri-plugins https://containers.github.io/nri-plugins
helm repo update nri-plugins
helm install -n kube-system balloons nri-plugins/nri-resource-policy-balloons
```

> **NOTE**: With containerd version earlier than v2.0 you need to enable
> the NRI support in the containerd configuration file. Instead of manual
> configuration you can provide `--set nri.runtime.patchConfig=true` to the Helm
> command above, which will automatically patch the containerd configuration
> file on each node.

Verify that the balloons policy plugin is running on every node:

```bash
$ kubectl -n kube-system get ds nri-resource-policy-balloons
NAME                           DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR            AGE
nri-resource-policy-balloons   2         2         2       2            2           kubernetes.io/os=linux   77s
```

## Configuration of Balloons Policy Plugin

The aim of the balloons policy configuration is to isolate the model (inference

> serving) containers to minimize the impact of containers on each other.

Author: Changed as sugggested

An example configuration for the current CPU-backed model profiles:

```yaml
cat <<EOF | kubectl apply -f -
apiVersion: config.nri/v1alpha1
kind: BalloonsPolicy
metadata:
  name: default
  namespace: kube-system
spec:
  allocatorTopologyBalancing: true
  balloonTypes:
  - name: kubeai-inference
    allocatorPriority: high
    preferNewBalloons: true
    hideHyperthreads: true
    matchExpressions:
    - key: labels/app.kubernetes.io/name
      operator: In
      values: ["vllm", "ollama"]
    - key: name
      operator: In
      values: ["server"]
    # Enables container-level information for the balloon in NodeResourceTopology objects
    # if agent.nodeResourceTopology is enabled.
    showContainersInNrt: true
  - name: default
    namespaces:
    - "*"
    shareIdleCPUsInSame: numa
  pinCPU: true
  pinMemory: false
  reservedPoolNamespaces:
  - kube-system
  reservedResources:
    cpu: "2"
  log:
    debug: ["policy"]
  instrumentation:
    httpEndpoint: :8891
    prometheusExport: true
  # NodeResourceTopology custom resource objects can be enabled for debugging
  # and closely # examining the CPU pinning of inference containers (not

>   # recommended for # production # because of potentially excessive load on the

Author: Changed/fixed

  # kube-apiserver).
  #agent:
  #  nodeResourceTopology: true
EOF
```

> The configuration above allocates full CPU cores to inference serving

Author: Check, changed as suggested

containers by hiding hyperthreads from them. For example, if a pod requests 6
CPUs, the balloon will reserve 6 full physical CPU cores (i.e., 12 logical
CPUs), but configure the cpuset so that the inference instance can only use 6

> logical CPUs—one per core—while the sibling hyperthreads remain unused. With
> this configuration for the `kubeai-inference` balloon, a new dedicated balloon
> is created for each inference engine container, as long as enough free CPUs are

Reviewer: Very long paragraph, split for readability:
```suggestion
logical CPUs—one per core—while the sibling hyperthreads remain unused.

Above `kubeai-inference` balloon configuration craetes a new dedicated balloon
for each inference engine container, as long as enough free CPUs are
```

Author: The paragraph is not super long. The sentences maybe. Reworded but still in one paragraph.
